### PR TITLE
Declare virtual attribute after 5.1

### PR DESCRIPTION
**Pull request by casperisfine**
Fix: #59 

Inspired from https://github.com/attr-encrypted/attr_encrypted/pull/263 which had similar deprecations. However  it's making some tests fail and I'm still unsure why :/

cc @rafaelfranca

### CODE REVIEW

**Comment by rafaelfranca** [comment]
I was discussing this deprecation with @sgrif today and asked him to take a look to see the best way to solve this problem. Using the `attribute` method looks sane, but last time I tried I could not figure it out how to fix the tests.

**Comment by sgrif** [comment]
I want to better understand why this difference in behavior from `ActiveRecord::Store` is needed. AR's just marks the store itself as changed (which can probably be removed). I'm curious if trying to mark the individual store attributes as changed was intentional or just a misreading of the AR code. I'm not sure that just calling `attribute` will be the right thing to do here.

**Comment by casperisfine** [comment]
> I'm curious if trying to mark the individual store attributes as changed was intentional

It is. There is tests for this: https://github.com/byroot/activerecord-typedstore/blob/154b5a8f37f77e80ac00a8f9591a1eb8c81f9e40/spec/active_record/typed_store_spec.rb#L48-L121

**Comment by sgrif** [comment]
> I want to better understand why this difference in behavior from ActiveRecord::Store is needed

**Comment by casperisfine** [comment]
Because the core idea behind this library is to have as little difference between those "virtual attributes" and attributes backed by their own DB column (short of querying obviously).

`ActiveRecord::Store` does very little in that regard.  No dirty tracking, no coercion, no `attribute?`, etc.

**Comment by sgrif** [comment]
I looked into when this was added. It appears my original assessment is correct, and this was a misreading of the old code. The `will_change!` line was added in https://github.com/byroot/activerecord-typedstore/commit/8d09313ee8caccd081431c3909ce1396ef02687a. The corresponding line which was removed was calling it on the store itself, not the sub-field (same as within Active Record). Since 4.2 detects mutations, the `will_change!` line can simply be removed.

**Comment by casperisfine** [comment]
> Since 4.2 detects mutations, the will_change! line can simply be removed.

Good to know, that will simplify code in the future. But for now the issue we're having is with 5.1

**Comment by sgrif** [comment]
5.1 is 3 releases after 4.2

**Comment by casperisfine** [comment]
🤦‍♂️ I read 5.2 because I tested your suggestion and it was causing failures to the test suite.

File: lib/active_record/typed_store/extension.rb [hunk 1]
```diff
@@ -45,6 +45,7 @@ def define_typed_store_attribute_methods
         return if @typed_store_attribute_methods_generated
         store_accessors.each do |attribute|
           define_attribute_method(attribute.to_s)
+          attribute(attribute.to_sym) if ActiveRecord.gem_version >= Gem::Version.new('5.1.0')
```
Comment (sgrif) [inline lib/active_record/typed_store/extension.rb]: This line should be removed as well.

**Comment by casperisfine** [comment]
So yeah, it gets rid of the warning, but now all the dirty tracking tests fail. I guess I'll have to dig deeper.

**Comment by sgrif** [comment]
Yeah I'm looking at it now, too. It looks like prior to this, we were manually defining the `_changed?` method -- I think that's the right thing to do here, for this to work properly (especially with mutation).

**Comment by rafaelfranca** [comment]
Fixed on master

**Pull request closed without merging**